pkg/providers/cloudstack/cloudstack.go

```diff
@@ -0,0 +1,765 @@
+	}
+	if _, ok := os.LookupEnv(eksaLicense); !ok {
+		if err := os.Setenv(eksaLicense, ""); err != nil {
+			return fmt.Errorf("unable to set %s: %v", eksaLicense, err)
+		}
+	}
+	return nil
+}
+
+func (p *cloudstackProvider) SetupAndValidateCreateCluster(ctx context.Context, clusterSpec *cluster.Spec) error {
+	err := p.validateEnv(ctx)
+	if err != nil {
+		return fmt.Errorf("failed setup and validations: %v", err)
+	}
+	if p.datacenterConfig.Spec.Insecure {
+		logger.Info("Warning: CloudStackDatacenterConfig configured in insecure mode")
+	}
+	if err := os.Setenv(cloudMonkeyInsecure, strconv.FormatBool(p.datacenterConfig.Spec.Insecure)); err != nil {
+		return fmt.Errorf("unable to set %s: %v", cloudMonkeyInsecure, err)
+	}
+
+	// TODO: Integrate with validator (defined in https://github.com/aws/eks-anywhere/pull/1256/)
+	_ = validateControlPlaneHost(&clusterSpec.Spec.ControlPlaneConfiguration.Endpoint.Host)
+
+	if err := p.setupSSHAuthKeysForCreate(); err != nil {
+		return fmt.Errorf("failed setup and validations: %v", err)
+	}
+
+	if clusterSpec.IsManaged() {
+		for _, mc := range p.MachineConfigs() {
+			em, err := p.providerKubectlClient.SearchCloudStackMachineConfig(ctx, mc.GetName(), clusterSpec.ManagementCluster.KubeconfigFile, mc.GetNamespace())
+			if err != nil {
+				return err
+			}
+			if len(em) > 0 {
+				return fmt.Errorf("CloudStackMachineConfig %s already exists", mc.GetName())
+			}
+		}
+		existingDatacenter, err := p.providerKubectlClient.SearchCloudStackDatacenterConfig(ctx, p.datacenterConfig.Name, clusterSpec.ManagementCluster.KubeconfigFile, clusterSpec.Namespace)
+		if err != nil {
+			return err
+		}
+		if len(existingDatacenter) > 0 {
+			return fmt.Errorf("CloudStackDeployment %s already exists", p.datacenterConfig.Name)
+		}
+	}
+	if p.skipIpCheck {
+		logger.Info("Skipping check for whether control plane ip is in use")
+		return nil
+	}
+
+	return nil
+}
+
+func (p *cloudstackProvider) SetupAndValidateUpgradeCluster(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+	return fmt.Errorf("upgrade is not yet supported for CloudStack cluster")
+}
+
+func (p *cloudstackProvider) SetupAndValidateDeleteCluster(ctx context.Context) error {
+	err := p.validateEnv(ctx)
+	if err != nil {
+		return fmt.Errorf("failed setup and validations: %v", err)
+	}
+	return nil
+}
+
+type CloudStackTemplateBuilder struct {
+	datacenterConfigSpec       *v1alpha1.CloudStackDatacenterConfigSpec
+	controlPlaneMachineSpec    *v1alpha1.CloudStackMachineConfigSpec
+	workerNodeGroupMachineSpec *v1alpha1.CloudStackMachineConfigSpec
+	etcdMachineSpec            *v1alpha1.CloudStackMachineConfigSpec
+	now                        types.NowFunc
+}
+
+func (vs *CloudStackTemplateBuilder) WorkerMachineTemplateName(clusterName string) string {
+	t := vs.now().UnixNano() / int64(time.Millisecond)
+	return fmt.Sprintf("%s-worker-node-template-%d", clusterName, t)
```

Comment:
This looks like logic that can be shared too?

Reply:
I thought the naming conventions were the same even without worker node group support?

Reply:
I think we can use the same naming convention but just not support multiple worker node groups. Unless you are saying we don't support worker nodes at all?

Reply:
Yep, that's what I've done in the latest couple commits